DSL/com.larsreimann.safeds/src/test/resources/validation/schema/inferInitialSchema.sdstest

```diff
@@ -0,0 +1,91 @@
+        constraint {
+            /*
+            Suggestion:
+            no "val" while assigning to a non intermediate/resultant Schema
+            remember the type parameter constraint uses similar syntax
+                T2 super Number
+
+            maybe special operater "::" like
+                InitialSchema :: $ReadSchema(datasetName, datasetPath)
+            */
+            InitialSchema = $ReadSchema(datasetName, datasetPath)
+        }
+    }
+
+    fun checkColumns <ComputedSchema :: $SchemaType> (vararg columnName: String) -> dataset: Dataset<ComputedSchema> {
+        constraint {
+            ComputedSchema = $CheckColumn(CurrentSchema, ColumnName)
+        }
+    }
+
+    fun joinDatasets <ComputedSchema :: $SchemaType, OtherDatasetSchema :: $SchemaType> (
+            otherDataset: Dataset<OtherDatasetSchema>,
+            joinOnColumn: String
+        ) -> dataset: Dataset<ComputedSchema> {
+        constraint {
+            /* here we access to OtherDatasetSchema and ComputedSchema schemas */
+            ComputedSchema = $JoinSchema(CurrentSchema, OtherDatasetSchema, joinOnColumn)
+        }
+    }
+}
+
+workflow myWorkflow {
+    val data = Dataset.loadDataset(datasetName, datasetPath);
+}
+
+/*
+## Concept
+
+- The main Idea is that "Schema"s are special "Types" declared in the type parameter with special "NAME :: $SchemaType" syntax which for
+  functions which declare it, forgoes the necessity to specify the Scheme Type upon function invocation, unlike usual
+  type parameter declared by functions
+    - the syntax
+      Dataset.loadDataset("datasetNameSTR", "datasetPathSTR");
+      insted of
+      Dataset.loadDataset<ASchemaType>(datasetName, datasetPath)
+
+- The Schema types are infered/computed automatically by the goals in the constraint
+
+- These Schemas types can be thought as more like custom names we are using to refer to the actuall internal Data schemas of the objects which
+  can have a Data schema associated with them
+
+## Rules for Classes that use a Schema Type type parameter like "<CurrentSchema :: $SchemaType>" in above example
+
+- Static functions do NOT have access to the Schema in class's type parameters.
+    - "loadDataset" can't refer to "CurrentSchema".
+
+- Non-static functions have access to the Schema in class's type parameters.
+    - "checkColumns" and "joinDatasets" refer to "CurrentSchema"
+
+- All the other (than the Schema defind in classes's type parameters) Schemas that any function refers to must be declared in type parameters
+  of the function. The parameters/results of functions refer to these Schema types in their type arguments (ofcourse if the parameters/results
+  are instances with Schema)
+    - function "joinDatasets" declares "ComputedSchema" and "OtherDatasetSchema" Schemas in it's type parameters. The parameter "otherDataset"
+      and result "dataset" have these Schemas respectively
```

Comment:
I think you mean the right thing, we just need to be careful with the terminology: We sometimes use "Schema" to refer to the type parameter (e.g. `CurrentSchema` in `<CurrentSchema :: $SchemaType>`) and sometimes the actual instances of a schema (e.g. `MySchema` in `schema MySchema`).

Reply:
Sure, we will call it Schema Type Parameter.

Reply:
Another idea, can we call the `MySchema` in `schema MySchema` "Initial Schema" instead. We can also change the keyword to "InitialSchema" so it will be `initialSchema MySchema`. It is a bit long for a keyword but it makes it clear what it really is also this allows us to call the "Schema type parameter", just "Schema".

Reply:
@lars-reimann  do give your feedback

Reply:
I think `schema MySchema`, i.e. what we have now, is fine. It's not necessarily the initial schema: We could also refer to some schema in a predicate.

Reply:
Okay